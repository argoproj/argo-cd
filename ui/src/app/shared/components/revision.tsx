import * as React from 'react';
import {revisionUrl} from './urls';

export const Revision = ({repoUrl, revision, children}: {repoUrl: string; revision: string; children?: React.ReactNode}) => {
    revision = revision || '';
    const url = revisionUrl(repoUrl, revision);
    const content = children || (isSHA(revision) ? revision.substr(0, 7) : revision);
    return url !== null ? (
        <a href={url} target='_blank' rel='noopener noreferrer'>
            {content}
        </a>
    ) : (
        <span>{content}</span>
    );
};

export const isSHA = (revision: string) => {
    // https://stackoverflow.com/questions/468370/a-regex-to-match-a-sha1
    return revision.match(/^[a-f0-9]{5,40}$/) !== null;
};
