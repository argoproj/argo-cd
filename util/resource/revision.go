package resource

import (
	"strconv"

	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
)

func GetRevision(obj *unstructured.Unstructured) int64 {
	if obj == nil {
		return 0
	}
	for _, name := range []string{"deployment.kubernetes.io/revision", "rollout.argoproj.io/revision"} {
		text, ok := obj.GetAnnotations()[name]
		if ok {
			revision, _ := strconv.ParseInt(text, 10, 64)
			return revision
		}
	}
	return 0
}
