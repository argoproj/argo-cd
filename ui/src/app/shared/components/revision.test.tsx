import * as renderer from 'react-test-renderer';
import * as React from 'react';
import {isSHA, Revision} from "./revision";

test('Revision.SHA1.Children', () => {
    const tree = renderer.create(<Revision repoUrl='http://github.com/my-org/my-repo' revision='24eb0b24099b2e9afff72558724e88125eaa0176'>foo</Revision>).toJSON();

    expect(tree).toMatchSnapshot()
});

test('Revision.SHA1.NoChildren', () => {
    const tree = renderer.create(<Revision repoUrl='http://github.com/my-org/my-repo' revision='24eb0b24099b2e9afff72558724e88125eaa0176'/>).toJSON();

    expect(tree).toMatchSnapshot()
});

test('Revision.Branch.Children', () => {
    const tree = renderer.create(<Revision repoUrl='http://github.com/my-org/my-repo' revision='long-branch-name'>foo</Revision>).toJSON();

    expect(tree).toMatchSnapshot()
});


test('Revision.Branch.NoChildren', () => {
    const tree = renderer.create(<Revision repoUrl='http://github.com/my-org/my-repo' revision='long-branch-name'/>).toJSON();

    expect(tree).toMatchSnapshot()
});

test('isSHA1', () => {
    expect(isSHA('24eb0b24099b2e9afff72558724e88125eaa0176')).toBe(true);
    expect(isSHA('master')).toBe(false);
});