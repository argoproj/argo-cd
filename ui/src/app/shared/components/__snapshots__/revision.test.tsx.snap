// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Revision.Branch.Children 1`] = `
<a
  href="http://github.com/my-org/my-repo/tree/long-branch-name"
  rel="noopener noreferrer"
  target="_blank"
>
  foo
</a>
`;

exports[`Revision.Branch.NoChildren 1`] = `
<a
  href="http://github.com/my-org/my-repo/tree/long-branch-name"
  rel="noopener noreferrer"
  target="_blank"
>
  long-branch-name
</a>
`;

exports[`Revision.SHA1.Children 1`] = `
<a
  href="http://github.com/my-org/my-repo/commit/24eb0b24099b2e9afff72558724e88125eaa0176"
  rel="noopener noreferrer"
  target="_blank"
>
  foo
</a>
`;

exports[`Revision.SHA1.NoChildren 1`] = `
<a
  href="http://github.com/my-org/my-repo/commit/24eb0b24099b2e9afff72558724e88125eaa0176"
  rel="noopener noreferrer"
  target="_blank"
>
  24eb0b2
</a>
`;
